rpcs3/Emu/Cell/lv2/sys_uart.cpp

```diff
@@ -1,35 +1,603 @@
+		return true;
+	}
+};
+
+struct generic_reply_cmd : public ps3av_cmd
+{
+	u16 get_size(vuart_av_thread &vuart, const void *pkt_buf) override
+	{
+		static_cast<void>(vuart);
+		static_cast<void>(pkt_buf);
+		return 0;
+	}
+
+	bool execute(vuart_av_thread &vuart, const void *pkt_buf, u32 reply_max_size) override
+	{
+		auto pkt = static_cast<const ps3av_header *>(pkt_buf);
+
+		vuart.write_resp(pkt->cid, PS3AV_STATUS_SUCCESS);
+		return true;
+	}
+};
+
+error_code sys_uart_initialize(ppu_thread &ppu)
+{
+	ppu.state += cpu_flag::wait;
+
+	sys_uart.trace("sys_uart_initialize()");
+
+	static shared_mutex m{};
+	std::lock_guard lock(m);
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (vuart_thread.init)
+	{
+		return CELL_EPERM;
+	}
+
+	vuart_thread.init = 1;
 
 	return CELL_OK;
 }
 
-error_code sys_uart_receive(vm::ptr<void> buffer, u64 size, u32 unk)
+error_code sys_uart_receive(ppu_thread &ppu, vm::ptr<void> buffer, u64 size, u32 mode)
 {
-	sys_uart.todo("sys_uart_receive(buffer=*0x%x, size=0x%llx, unk=0x%x)", buffer, size, unk);
+	sys_uart.trace("sys_uart_receive(buffer=*0x%x, size=0x%llx, mode=0x%x)", buffer, size, mode);
 
-	return CELL_OK;
+	if (!size)
+	{
+		return CELL_OK;
+	}
+
+	if ((mode & ~1UL) != 0)
+	{
+		return CELL_EINVAL;
+	}
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (!vuart_thread.init)
+	{
+		return CELL_ESRCH;
+	}
+
+	if (size > 0x20000U)
+	{
+		// kmalloc restriction
+		fmt::throw_exception("Buffer is too big");
+	}
+
+	if (mode > 1)
+	{
+		return CELL_EINVAL;
+	}
+
+	const std::unique_ptr<u8[]> data = std::make_unique<u8[]>(size);
+	u32 read_size = 0;
+
+	auto vuart_read = [&](u8 *buf, u32 buf_size) -> s32
+	{
+		const u32 ITER_SIZE = 4096;
+		std::unique_lock lock(vuart_thread.rx_mutex, std::defer_lock);
+
+		if (!lock.try_lock())
+		{
+			return CELL_EBUSY;
+		}
+
+		const u32 first_iter_size = std::min(ITER_SIZE, buf_size);
+		const u32 first_iter_read = vuart_thread.read_rx_data(buf, first_iter_size);
+
+		if (first_iter_read < first_iter_size || buf_size <= ITER_SIZE)
+		{
+			return first_iter_read;
+		}
+
+		u32 read_size;
+
+		for (read_size = ITER_SIZE; read_size + ITER_SIZE < buf_size; read_size += ITER_SIZE)
+		{
+			const u32 in_loop_read = vuart_thread.read_rx_data(buf + read_size, ITER_SIZE);
+
+			if (in_loop_read < ITER_SIZE)
+			{
+				return read_size + in_loop_read;
+			}
+		}
+
+		return read_size + vuart_thread.read_rx_data(buf + read_size, buf_size - read_size);
+	};
+
+	if (mode == 1)
+	{
+		// Yield before checking for packets
+		lv2_obj::sleep(ppu);
+
+		for (;;)
+		{
+			if (ppu.is_stopped())
+			{
+				return {};
+			}
+
+			std::unique_lock<shared_mutex> lock(vuart_thread.rx_wake_m);
+			const s32 read_result = vuart_read(data.get(), static_cast<u32>(size));
+
+			if (read_result > CELL_OK)
+			{
+				read_size = read_result;
+				break;
+			}
+
+			vuart_thread.rx_wake_c.wait_unlock(5000, lock);
+		}
+
+		if (ppu.check_state())
```

Comment:
The new coding model compatible with savestates is to avoid early return as much as possible, even when we know emulation is stopping. check_state can be executed without if here. (the is_stopped() before is valid for savestates though)